File: functions/python/daily_sentence.py
"""毎日例文の配信ロジック（純粋関数部分）。

Cloud Functions のエントリポイントは daily_sentence_handlers.py にある。
テスト容易性のため、判定ロジックはここに副作用なしで置く。
"""

from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo

# 段階ごとの配信間隔（日）。notify_tier == len(...) で配信停止。
TIER_INTERVAL_DAYS = [1, 3, 10, 30]
TIER_STOPPED = len(TIER_INTERVAL_DAYS)
# 同じ段階で無反応がこの回数たまると次の段階へ落とす
TIER_MAX_MISSES = 3

DEFAULT_TIMEZONE = "Asia/Tokyo"
DEFAULT_GENERATION_HOUR = 10

# キャッシュヒットするまでターゲット語を引き直す回数
MAX_TARGET_WORD_RETRY = 5


def _local(tz_name: str | None, now: datetime) -> datetime:
    """不正・未設定のtz名は Asia/Tokyo にフォールバックしてローカル時刻へ変換する。"""
    try:
        tz = ZoneInfo(tz_name or DEFAULT_TIMEZONE)
    except Exception:
        tz = ZoneInfo(DEFAULT_TIMEZONE)
    return now.astimezone(tz)


def local_hour(tz_name: str | None, now: datetime) -> int:
    return _local(tz_name, now).hour


def local_date(tz_name: str | None, now: datetime) -> str:
    return _local(tz_name, now).strftime("%Y-%m-%d")


def notify_utc_hour(user_data: dict, now: datetime) -> int | None:
    """現地の配信希望時刻が UTC の何時の起動に当たるかを求める。

    users/{uid}.notify_utc_hour に非正規化して配信対象クエリの絞り込みに使う
    （毎時 users 全件を読むのを避けるため）。値の維持は主に dailyBatch が行うが、
    クライアントが設定を書いた直後に反映させるためロジックはここにも置く。

    オフセットの引き算ではなく24通りを実際にローカル変換して探す。+5:30 / +5:45 の
    ような分単位オフセットでも現地の各時刻はUTCのいずれか1時刻と1対1に対応する
    （配信が現地 10:30 になるズレはあるが、これは従来の挙動と同じ）。

    DST 春の切り替え日はその現地時刻自体が存在しないため None を返す。
    """
    preferred = user_data.get("preferred_generation_hour", DEFAULT_GENERATION_HOUR)
    day = now.astimezone(timezone.utc).replace(minute=0, second=0, microsecond=0)
    for hour in range(24):
        if local_hour(user_data.get("timezone"), day.replace(hour=hour)) == preferred:
            return hour
    return None


def _as_datetime(value) -> datetime | None:
    """Firestore の timestamp を tz-aware datetime に正規化する。"""
    if value is None:
        return None
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    return None


def has_generation_history(user_data: dict) -> bool:
    """一度でも例文を生成したことがあるか。

    last_sentence_generated_at が導入される前に生成したきりのユーザーを
    estimated_vocab で救う。
    """
    if user_data.get("last_sentence_generated_at") is not None:
        return True
    return user_data.get("estimated_vocab", 0) > 0


def is_due(user_data: dict, now: datetime) -> bool:
    """段階ごとの配信間隔を満たしているか。初回（未通知）は常に True。"""
    tier = user_data.get("notify_tier", 0)
    if tier >= TIER_STOPPED:
        return False
    last_notified = _as_datetime(user_data.get("last_notified_at"))
    if last_notified is None:
        return True
    return now >= last_notified + timedelta(days=TIER_INTERVAL_DAYS[tier])


def evaluate_response(user_data: dict) -> dict:
    """前回通知への反応を評価し、更新後の段階を返す。

    「次へ」押下（last_sentence_generated_at）を主シグナル、開封（last_opened_at）を
    副シグナルとする。どちらも動いていない場合だけ段階が進む。

    開封は段階（notify_tier）までは戻さないが、無反応の連続カウントは0に戻す。
    届いた通知に反応している以上、間隔を広げる方向へ進めるべきではないため。
    """
    tier = user_data.get("notify_tier", 0)
    misses = user_data.get("notify_tier_misses", 0)
    last_notified = _as_datetime(user_data.get("last_notified_at"))

    if last_notified is None:
        return {"notify_tier": tier, "notify_tier_misses": misses}

    generated_at = _as_datetime(user_data.get("last_sentence_generated_at"))
    if generated_at is not None and generated_at > last_notified:
        return {"notify_tier": 0, "notify_tier_misses": 0}

    opened_at = _as_datetime(user_data.get("last_opened_at"))
    if opened_at is not None and opened_at > last_notified:
        return {"notify_tier": tier, "notify_tier_misses": 0}

    misses += 1
    if misses >= TIER_MAX_MISSES:
        tier += 1
        misses = 0
    return {"notify_tier": tier, "notify_tier_misses": misses}


def uses_premium_trial(user_data: dict, now: datetime | None = None) -> bool:
    """この配信をプレミアム体験トライアル枠（premium品質）で出すか。

    トライアルは期間制なので、期間中の free ユーザーは配信も premium 品質にする。
    判定は generateThaiSentence 側（sentence_handlers._resolve_trial_active）と同じ
    期限のみで、消費という概念は無い。
    """
    if user_data.get("tier") == "premium":
        return False
    expires_at = user_data.get("premium_trial_expires_at")
    if not isinstance(expires_at, datetime):
        return False
    if expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=timezone.utc)
    return (now or datetime.now(timezone.utc)) < expires_at


def delivery_skip_reason(user_data: dict, now: datetime) -> str | None:
    """配信を見送る理由。配信対象なら None。

    理由を文字列で返すのは、配信バッチのログに内訳を残すため。
    「通知が届いていない」を調べるたびに Firestore を読んで条件を手で再現するのは
    非効率なので、判断に使う値は常設ログにしておく。
    """
    if not has_generation_history(user_data):
        return "no_history"
    if user_data.get("daily_reminder_enabled") is False:
        return "opt_out"
    if not user_data.get("fcm_token"):
        return "no_token"
    if user_data.get("daily_sentence_generated"):
        return "already_generated"
    if user_data.get("remaining_sentences", 0) <= 0:
        return "quota_exhausted"
    if user_data.get("notify_tier", 0) >= TIER_STOPPED:
        return "backoff_stopped"

    hour = local_hour(user_data.get("timezone"), now)
    if hour != user_data.get("preferred_generation_hour", DEFAULT_GENERATION_HOUR):
        # notify_utc_hour で絞ってから呼ぶので、これが出るのは非正規化値が古いとき。
        return "hour_mismatch"

    if not is_due(user_data, now):
        return "not_due"
    return None


def should_deliver(user_data: dict, now: datetime) -> bool:
    """配信対象かどうか。反応評価の前に効く足切り条件をまとめて判定する。"""
    return delivery_skip_reason(user_data, now) is None

File: functions/go/internal/dailysentence/dailysentence.go
// Package dailysentence は毎日例文の配信判定（純粋ロジック）。
// functions/python/daily_sentence.py の移植。
//
// Cloud Functions のエントリポイントは別（daily_sentence.go）。
// テスト容易性のため、判定ロジックはここに副作用なしで置く。
package dailysentence

import (
	"strconv"
	"strings"
	"time"

	"github.com/mnbst/thai-memo/functions/go/internal/notify"
	"github.com/mnbst/thai-memo/functions/go/internal/sentence"
)

// TierIntervalDays は段階ごとの配信間隔（日）。
// notify_tier == len(TierIntervalDays) で配信停止。
var TierIntervalDays = []int{1, 3, 10, 30}

// TierStopped は配信停止に達した段階。
var TierStopped = len(TierIntervalDays)

const (
	// TierMaxMisses は同じ段階で無反応がこの回数たまると次の段階へ落とす。
	TierMaxMisses = 3

	// DefaultTimezone はタイムゾーン未設定・不正時のフォールバック。
	DefaultTimezone = "Asia/Tokyo"

	// DefaultGenerationHour は配信希望時刻のデフォルト。
	DefaultGenerationHour = 10

	// MaxTargetWordRetry はキャッシュヒットするまでターゲット語を引き直す回数。
	MaxTargetWordRetry = 5
)

// local は不正・未設定の tz 名を Asia/Tokyo にフォールバックしてローカル時刻へ変換する。
func local(tzName any, now time.Time) time.Time {
	name, ok := tzName.(string)
	if !ok || name == "" {
		name = DefaultTimezone
	}
	loc, err := time.LoadLocation(name)
	if err != nil {
		loc, err = time.LoadLocation(DefaultTimezone)
		if err != nil {
			return now.UTC()
		}
	}
	return now.In(loc)
}

// LocalHour は現地時刻の「時」。
func LocalHour(tzName any, now time.Time) int {
	return local(tzName, now).Hour()
}

// LocalDate は現地時刻の日付を "YYYY-MM-DD" で返す。
func LocalDate(tzName any, now time.Time) string {
	return local(tzName, now).Format("2006-01-02")
}

// NotifyUTCHour は現地の配信希望時刻が UTC の何時の起動に当たるかを求める。
//
// JS 版（utils/notifyUtcHour.ts）と同じ計算で、実装は internal/notify に集約している。
// Python 側とも同じ結果になることは internal/notify の差分テストで確認済み。
func NotifyUTCHour(userData map[string]any, now time.Time) (int, bool) {
	return notify.UTCHour(
		userData["timezone"], userData["preferred_generation_hour"], now)
}

// asDatetime は Firestore の timestamp を time.Time に正規化する。
func asDatetime(value any) (time.Time, bool) {
	t, ok := value.(time.Time)
	return t, ok
}

// numField は Firestore の数値を int にする。整数は int64 で返るため。
func numField(value any, fallback int) int {
	switch n := value.(type) {
	case int64:
		return int(n)
	case float64:
		return int(n)
	case int:
		return n
	}
	return fallback
}

// HasGenerationHistory は一度でも例文を生成したことがあるか。
//
// last_sentence_generated_at が導入される前に生成したきりのユーザーを
// estimated_vocab で救う。
func HasGenerationHistory(userData map[string]any) bool {
	if userData["last_sentence_generated_at"] != nil {
		return true
	}
	return numField(userData["estimated_vocab"], 0) > 0
}

// IsDue は段階ごとの配信間隔を満たしているか。初回（未通知）は常に true。
func IsDue(userData map[string]any, now time.Time) bool {
	tier := numField(userData["notify_tier"], 0)
	if tier >= TierStopped {
		return false
	}
	lastNotified, ok := asDatetime(userData["last_notified_at"])
	if !ok {
		return true
	}
	due := lastNotified.AddDate(0, 0, TierIntervalDays[tier])
	return !now.Before(due)
}

// TierUpdate は反応評価の結果（Firestore へ書く内容）。
type TierUpdate struct {
	NotifyTier   int
	NotifyMisses int
}

// EvaluateResponse は前回通知への反応を評価し、更新後の段階を返す。
//
// 「次へ」押下（last_sentence_generated_at）を主シグナル、開封（last_opened_at）を
// 副シグナルとする。どちらも動いていない場合だけ段階が進む。
//
// 開封は段階（notify_tier）までは戻さないが、無反応の連続カウントは0に戻す。
// 届いた通知に反応している以上、間隔を広げる方向へ進めるべきではないため。
func EvaluateResponse(userData map[string]any) TierUpdate {
	tier := numField(userData["notify_tier"], 0)
	misses := numField(userData["notify_tier_misses"], 0)

	lastNotified, ok := asDatetime(userData["last_notified_at"])
	if !ok {
		return TierUpdate{NotifyTier: tier, NotifyMisses: misses}
	}

	if generatedAt, ok := asDatetime(userData["last_sentence_generated_at"]); ok &&
		generatedAt.After(lastNotified) {
		return TierUpdate{NotifyTier: 0, NotifyMisses: 0}
	}

	if openedAt, ok := asDatetime(userData["last_opened_at"]); ok &&
		openedAt.After(lastNotified) {
		return TierUpdate{NotifyTier: tier, NotifyMisses: 0}
	}

	misses++
	if misses >= TierMaxMisses {
		tier++
		misses = 0
	}
	return TierUpdate{NotifyTier: tier, NotifyMisses: misses}
}

// UsesPremiumTrial はこの配信をプレミアム体験トライアル枠（premium品質）で出すか。
//
// トライアルは期間制なので、期間中の free ユーザーは配信も premium 品質にする。
// 判定は generateThaiSentence 側と同じ期限のみで、消費という概念は無い。
func UsesPremiumTrial(userData map[string]any, now time.Time) bool {
	if userData["tier"] == "premium" {
		return false
	}
	expiresAt, ok := asDatetime(userData["premium_trial_expires_at"])
	if !ok {
		return false
	}
	return now.Before(expiresAt)
}

// DeliverySkipReason は配信を見送る理由。配信対象なら空文字。
//
// 理由を文字列で返すのは、配信バッチのログに内訳を残すため。
// 「通知が届いていない」を調べるたびに Firestore を読んで条件を手で再現するのは
// 非効率なので、判断に使う値は常設ログにしておく。
func DeliverySkipReason(userData map[string]any, now time.Time) string {
	if !HasGenerationHistory(userData) {
		return "no_history"
	}
	// Python の `is False`。未設定は対象外にしない。
	if enabled, ok := userData["daily_reminder_enabled"].(bool); ok && !enabled {
		return "opt_out"
	}
	if token, _ := userData["fcm_token"].(string); token == "" {
		return "no_token"
	}
	if generated, ok := userData["daily_sentence_generated"].(bool); ok && generated {
		return "already_generated"
	}
	if numField(userData["remaining_sentences"], 0) <= 0 {
		return "quota_exhausted"
	}
	if numField(userData["notify_tier"], 0) >= TierStopped {
		return "backoff_stopped"
	}

	hour := LocalHour(userData["timezone"], now)
	if hour != numField(userData["preferred_generation_hour"], DefaultGenerationHour) {
		// notify_utc_hour で絞ってから呼ぶので、これが出るのは非正規化値が古いとき。
		return "hour_mismatch"
	}

	if !IsDue(userData, now) {
		return "not_due"
	}
	return ""
}

// ShouldDeliver は配信対象かどうか。
func ShouldDeliver(userData map[string]any, now time.Time) bool {
	return DeliverySkipReason(userData, now) == ""
}

// DailyBatchSize は1回の配信で作る例文の本数。
//
// 例文→確認クイズ→まとめクイズのサイクルを1日で1周させるため、
// アプリからの生成（generateThaiSentence）と同じ「1セット」の本数を使う。
const DailyBatchSize = sentence.SetSize

// DailyBatchMinVersion は5本セットを扱えるクライアントの最小バージョン。
//
// 旧版は配信済みドキュメントを全部取り込んだうえで最新の1件しか表示しないため、
// 5本送るとクォータだけ消費して4本が履歴に埋もれる。必ず版で絞る。
// ロールアウトはこの値の上げ下げで切り替える。
var DailyBatchMinVersion = [3]int{1, 4, 8}

// BatchSize は配信本数。版が読めない・古い場合は従来どおり1本（安全側）。
//
// app_build_number は使えない。pubspec のビルド番号は 0 固定で、実際の番号は
// CI が --build-number=${{ github.run_number }} で注入しており、tester と prod で
// ワークフローが別＝採番系列が独立しているため大小がリリース順序を表さない。
func BatchSize(userData map[string]any) int {
	version, _ := userData["app_version"].(string)
	if compareVersion(version, DailyBatchMinVersion) < 0 {
		return 1
	}
	return DailyBatchSize
}

// compareVersion は "1.4.8" 形式を major/minor/patch の順に比較する。
// パースできない値は最小扱い（-1）にして安全側へ倒す。
func compareVersion(version string, other [3]int) int {
	parts := strings.SplitN(strings.TrimSpace(version), ".", 4)
	if len(parts) != 3 {
		return -1
	}
	for i, p := range parts {
		n, err := strconv.Atoi(p)
		if err != nil || n < 0 {
			return -1
		}
		if n != other[i] {
			if n < other[i] {
				return -1
			}
			return 1
		}
	}
	return 0
}
